### ISSUE TITLE
CreateInstance called on abstract types

### ISSUE THREAD

**Issue by blairconrad** [FakeItEasy/FakeItEasy#494]
As raised by @hshilling in https://github.com/FakeItEasy/FakeItEasy/issues/33#issuecomment-110595919, `TypeCatalogueInstanceProvider.InstantiateAllOfType<T>` attempts to instantiate found types (such as `IDummyDefinition`) without regard to whether the type is instantiable.
There's a blanket catch:

```csharp
try
{
    result.Add((T)Activator.CreateInstance(type));
}
catch
{
}
```

But in some cases, notably when debugging, the exception is killing the test runs.

While this may not fix all instances of #33 (it's not clear that all reports of the failure to create an instance of an interface are coming from the same source), and exceptions still may be thrown from this sort of code, I think we should avoid trying to instantiate abstract types when we can know better.

**Comment by blairconrad** [comment]
Sorry, should've said "as raised by @hshilling in https://github.com/FakeItEasy/FakeItEasy/issues/33#issuecomment-110595919".

**Comment by adamralph** [comment]
@blairconrad 

> But in some cases, notably when debugging, the exception is killing the test runs.

What is the effect on the test runs? I'm nitpicking, but I'm wondering if this is really a bug.

**Comment by blairconrad** [comment]
@adamralph you, nitpick? Hard to fathom. :grin:

It is a fair question. To quote @hshilling, who can perhaps expand:

> Below is the stack where it died (dies on try to create FakeItEasy.IDummyFactory):

Certainly when "break on thrown exceptions" is turned on, the debugger breaks on the thrown exception, which is not catastrophic, but at least irritating.

I think that knowingly causing an exception to be thrown in this case, _even though we catch them_ is an error in logic. It's true, though, that in many cases it's not visible to the clients, so I'm not against reclassifying, but at the time it didn't feel like a refactor or an enhancement, although arguments could be made for any of the 3 labels, I think. For example, maybe we're enhancing FIE by providing an optimization.

**Comment by blairconrad** [comment]
I should've said earlier, but the fix was release in 2.0.0-beta10.
I've retroactively added you name to the release notes, @hshilling!

**Comment by blairconrad** [comment]
This issue has been fixed in [release 2.0.0](https://github.com/FakeItEasy/FakeItEasy/releases/tag/2.0.0).
